app/src/main/java/com/epfl/drawyourpath/utils/Utils.kt

```diff
@@ -392,4 +394,118 @@ object Utils {
     fun coordinatesToBitmap(coordinates: List<LatLng>, size: Int = 100, paint: Paint = defaultPaint): Bitmap {
         return strokesToBitmap(listOf(coordinatesToStroke(coordinates)), size, paint)
     }
+
+    /**
+     * Reduces the number of points in a Path.
+     * @param path The path we want to reduce
+     * @param maxError The max error percentage relative to the path distance
+     * @return A similar Path with the least useful points removed.
+     */
+    fun reducePath(path: Path, maxError: Float = 0.01F): Path {
+        val reducedPoints = mutableListOf(listOf<LatLng>())
+        if (path.size() <= 2) {
+            return Path(path.getPoints())
+        }
+
+        val epsilon = maxError * path.getDistance().toFloat()
+        for (section in path.getPoints()) {
+            reducedPoints.add(reduceSection(section, epsilon))
+        }
+        reducedPoints.removeFirst()
+        return Path(reducedPoints.toList())
+    }
+
+    /**
+     * Reduces the number of points in a section.
+     * @param section The section we want to reduce
+     * @param epsilon The min distance to keep the point on the section
+     * @return A similar segment with the least useful points removed.
+     */
+    fun reduceSection(section: List<LatLng>, epsilon: Float): List<LatLng> {
+        val reducedPointList = mutableListOf<LatLng>()
+        if (section.size <= 2) {
+            reducedPointList.addAll(section)
+            return reducedPointList.toList()
+        }
+
+        val distances = distancesToSegment(section.subList(1, section.size - 1), section.first(), section.last())
+        val distMax = distances.max()
+        val indexMax = distances.indexOf(distMax) + 1
+        // Check if finished, otherwise we need to solve recursively
+        if (distMax <= epsilon) {
+            reducedPointList.add(section.first())
+            reducedPointList.add(section.last())
+        } else {
+            val firstHalf = reduceSection(section.subList(0, indexMax + 1), epsilon)
+            val secondHalf = reduceSection(section.subList(indexMax, section.size), epsilon)
+            reducedPointList.addAll(firstHalf.subList(0, firstHalf.size - 1))
+            reducedPointList.addAll(secondHalf)
+        }
+        return reducedPointList.toList()
```

Comment:
Clean recursive algorithm ! Do you think there is a risk that the algorith has to do a very deep recursion that could cause memory issues ?

Reply:
It is a neat algorithm ! For the performance the algorithm is $\Theta(n^2)$ and in the ideal case (when the recursion splits the work equally into two subtasks) it is $O(n*log(n))$, where n is the number of points in the original path.

The depth of the recursion should not be an issue : 
If we call m the number of segment in the reduced path then the depth of the recursion is
- log(m) in the ideal case
- m in the worst case 

So if we have a very long path that cannot be reduced well (like a very long zigzag) and that we mess with the epsilon parameter (the `reducePath` function should compensate with the length) then yes, but if you use the functions normally there should not be any issues whatsoever.

Reply:
Alright thanks for the clear answer. If it's not too risky then it's fine for me!